### DISCUSSION TITLE
Implement Type Hints for Python code

### DISCUSSION THREAD

**Discussion by loicteixeira** [wagtail/wagtail#14104] [Ideas]
As discussed in the core team meeting, we're going to start using Type Hints for Python code (and TypeScript for JS code, see the sister issue #5799).

This should improve the developer experience by adding type-checking (allowing us to catch issues as early as possible) and should improve the overall readability of the code.

Since it would not be realistic to add type hints to all the code base at once, it should not be enforced at first but only produce warning in the CI.

It has been activated on [draftjs_exporter](https://github.com/springload/draftjs_exporter) (a dependency of Wagtail), so it could provide some base configuration and stubs. Also, @thibaudcolas wrote [Python static type checking: field test](https://thib.me/python-static-type-checking-field-test), which goes onto the caveats and pitfall.

**Comment by AdrienLemaire** [comment]
@loicteixeira are updates for type hints implementation and coverage posted in this issue or somewhere else?

**Comment by loicteixeira** [comment]
Updates should be posted here yes, but as you can see, there hasn't been any progress so far as this isn't a critical issue.

**Comment by brylie** [comment]
I want to help with this issue. What would be a good place to start? For context, I'm adding type hints to a large Wagtail project I've built for around five years. It would give me a better feeling of confidence, if I could use mypy with Wagtail-related code.

**Comment by AbhishekCS3459** [comment]
Hey @loicteixeira I would like to help for the type checking in the js,form where should I start?

**Comment by jorenham** [comment]
@brylie Wagtail is very much depedent on Django, and Django isn't typed itself. To my knowledge, there are no "official" stubs package for Django. This means that official wagtail stubs are probably out of the question too. So I believe that the only feasible option is to build upon one of the unofficial Django type stubs packages, e.g. [django-stubs](https://github.com/typeddjango/django-stubs) (most popular, but last time I checked, it had several serieous issues), and [django-types](https://github.com/sbdchd/django-types) (a fork of django-stubs, addressing said issues). 
But this also goes to show how  creating a stubs package is a difficult job, especially for highly dynamic code, such as the django ORM, where types are created in runtime, and model fields are implemented in a very hacky way, in the pre-descriptor era.

As much as I would love to see a type-annotated wagtail, I also think that it'd be one of the most difficult projects to annotate (fragmented community opinions about typing, spaghetti code, a huge codebase that's changing faster than most people can keep up with, etc etc), I can imagine a team of experienced wagtail developers with solid python-typing experience making this possible. I just don't think that this should be considered a fun side project to learn about typing and wagtail; you'll only frustrate yourself (speaking from experience).

Having said that, I'm sure that there are places in the wagtail codebase that don't touch django too much, that aren't as difficult to annotate. It would require some very selective type checking configuration. 

Badly typed projects are worse than untyped ones. And having to spend days on solving typing issues caused by a depdendency is very frsutrating :)

**Comment by caipirginka** [comment]
I've come here due to a PyLance error in VS Code with Python 3.13 and Wagtail 7.2.1:
```
Cannot access attribute "live" for class "BaseManager[NewsPage]"
  Attribute "live" is unknown
```
at a line such as:
`NewsPage.objects.live()`

seemingly related to the missing Type Hints.
Claude Code solved it by creating a Protocol that can be used in all Page derived classes for their "objects" Django manager:
```
class WagtailPageManager(Protocol):
    """Protocol for Wagtail's PageManager methods.

    Wagtail doesn't provide type stubs, so this Protocol defines the interface
    for Wagtail's custom manager methods that are not recognized by type checkers.

    Use this to annotate the `objects` manager on Page models:

    Example:
        from typing import ClassVar
        from webcommon.typing import WagtailPageManager

        class MyPage(BasePage):
            objects: ClassVar[WagtailPageManager]

    See: https://github.com/wagtail/wagtail/issues/5803
    """
```
I attach it here the entire class, in case it's of any interest:

[typing.py](https://github.com/user-attachments/files/24803596/typing.py)

otherwise, please pardon me.

**Comment by jorenham** [comment]
> Claude Code solved it

I wouldn't consider that a solution, but rather a (very) convoluted workaround; a `# type: ignore` would've also done the trick.

**Comment by caipirginka** [comment]
I consider `# type: ignore` a (not very) convoluted workaround... 😅
In fact I always tell AI agents to avoid it and other similar solutions, if at all possible.
In the end, one must simply remember to use:
`objects: ClassVar[WagtailPageManager]`
just **once** per Page class, instead that ignoring **every** single method usage, so I prefer this,
but I guess it's just a matter of taste...

**Comment by thibaudcolas** [comment]
I’m moving this to discussions now as I think this is a very open-ended task that is unlikely to be tackled as a single issue. We have some type hints in Python code currently, if we want to increase our type coverage we need to break this up into smaller tasks that are more approachable for contributors. 

For people who want to help, here are some of the things I think we’d need:

- Here in comments, feedback on:
	- What parts of Wagtail to type sooner than later and to what degree.
	- Which recent (3.10+) Python typing features which could greatly simplify our adoption.
	- Type checker configuration recommendations (beyond just "mypy vs. pyright vs. ty")
	- Any other insight from retroactively adopting type hints on comparable projects (large **packages** used by others’ projects)
- In our issue tracker:
	- A proposal on how we’d update our tooling / docs to support this adoption of type hints (which tools we use, configured how, how to document our expectations for contributors)
	- Requests to type specific parts of Wagtail where it’s pretty clear that it’s very feasible. For example image utilities definitely feasible, the ORM not so much.
- In separate projects:
	- Open source examples of Wagtail sites that are typed
	- Open source examples of Django or Wagtail packages that are typed

I think someone creating a [Wagtail RFC](https://github.com/wagtail/rfcs) with a top-down plan on what we need to do would also help. We’re pretty clear we want more type hints, but it’s a big task so it’s well worth the planning work upfront. See for example our [RFC 78 for Stimulus adoption](https://github.com/wagtail/rfcs/blob/main/text/078-adopt-stimulus-js.md).

**Reply by jorenham** [reply]
Great to hear that there's growing interest for static typing here!

> * What parts of Wagtail to type sooner than later and to what degree.

I'd put the `Page` model high on the priority list.

> * Which recent (3.10+) Python typing features which could greatly simplify our adoption.

In my experience this isn't something worth planning out beforehand. Because in practice, you just go with the typing features that are most appropriate for the task at hand.

> * Type checker configuration recommendations (beyond just "mypy vs. pyright vs. ty")

I'm clearly biased (I'm the author), but I think the scipy-stubs type-checker configs (currently mypy, basedpyright, pyrefly, and ty) is the way to go: https://github.com/scipy/scipy-stubs/blob/master/pyproject.toml

> * Any other insight from retroactively adopting type hints on comparable projects (large **packages** used by others’ projects)

As I mentioned, I wrote [scipy-stubs](https://github.com/scipy/scipy-stubs). It currently counts 90k lines of effective code (according to [scc](https://github.com/boyter/scc)), so I suppose that'd count as "large".
But tbh, the amount of insights I've gained through this can probably fill an entire book. Many of those are probably specific to numpy and scipy, so I'm not sure how applicable that would be here. 

One general thing I'd like to recommend, is to use a TDD approach. As in; write simple type-tests that use `typing.assert_type` to validate that type-checkers infer the types you want them to infer.

**Reply by thibaudcolas** [reply]
thank you ❤️ I have a few follow-up questions if that’s ok

> Because in practice, you just go with the typing features that are most appropriate for the task at hand

That matches my experience with more mature type systems, but the reason I raised that is Python’s typing story is very much evolving? For example TypedDict was a big addition, and for Python 3.11+, [PEP 655 – Marking individual TypedDict items as required or potentially-missing](https://peps.python.org/pep-0655/) seems like another big one, where I’d rather wait for us to raise our support targets so we get to do this once with the better semantics. Does that match your experience?

> I think the scipy-stubs type-checker configs (currently mypy, basedpyright, pyrefly, and ty)

Is this just a matter of testing against multiple popular real-world checkers, or are there other benefits to this? I’ve used mypy + ty on one project and even with just two, found that they disagreed on correct annotations often enough, in ways where it wasn’t necessarily obvious which was "right".

----

And last question as someone with no knowledge of the SciPy ecosystem – on [Why is scipy-stubs a separate package?](https://github.com/scipy/scipy-stubs#why-is-scipy-stubs-a-separate-package), what’s your thoughts on how that rationale transfers to other projects?

For Wagtail, I would have expected it’d be better for us to add type hints directly in our code rather than stubs. So we can see the benefits of type hints when working on our own code. So there’s no risk of the annotations falling behind?

**Reply by jorenham** [reply]
> I’d rather wait for us to raise our support targets so we get to do this once with the better semantics. Does that match your experience?

`if TYPE_CHECKING` blocks and in `.pyi` you can freely use the `typing_extensions` backports on older Python versions, even if `typing_extensions` is not installed (that's because the typeshed stdlib stubs include `typing_extensions` ([src](https://github.com/python/typeshed/blob/main/stdlib/typing_extensions.pyi)))

> Is this just a matter of testing against multiple popular real-world checkers, or are there other benefits to this?

There are two main reasons:

- All type-checkers have blind-spots, so it's good to have multiple sets of eyes looking for mistakes
- There are subtle differences in how type-checkers deal with e.g. overloads. By using multiple type-checkers you'll be able to spot those differences, and can adjust the code or stubs to minimize those differences

> And last question as someone with no knowledge of the SciPy ecosystem – on [Why is scipy-stubs a separate package?](https://github.com/scipy/scipy-stubs#why-is-scipy-stubs-a-separate-package), what’s your thoughts on how that rationale transfers to other projects?
> 
> For Wagtail, I would have expected it’d be better for us to add type hints directly in our code rather than stubs. So we can see the benefits of type hints when working on our own code. So there’s no risk of the annotations falling behind?

I guess that depends on how willing the wagtail maintainers are to e.g. require that new contributions are annotated. I know that in in the SciPy community there are many contributors that wouldn't be happy with that, so if we were to require that there, then we risk losing valuable contributors. So perhaps that's something that you could ask the community first, and then decide on whether you want to use a separate `wagtail-stubs` or bundle it directly with wagtail.
Usually the recommendation is to prefer inline annotations over stubs. Because that way, runtime type-checkers (like [beartype](https://github.com/beartype/beartype)) can also be used.

**Comment by jams2** [comment]
I recently discovered https://gitlab.com/thelabnyc/wagtail-stubs, which is a great effort but is not complete. In a quick test I noticed missing streamfield block metaclass attributes.

**Reply by jorenham** [reply]
See https://gitlab.com/thelabnyc/wagtail-stubs/-/blob/master/stubtest-allowlist.txt?ref_type=heads for a list of other missing/incorrect annotations.
